tests/app/celery/v3/test_notification_tasks.py

```diff
@@ -0,0 +1,240 @@
+    notification = notify_db_session.session.execute(query).one().Notification
+    assert notification.status == NOTIFICATION_PERMANENT_FAILURE
+    assert notification.status_reason == "The template does not exist."
+
+
+@pytest.mark.xfail(reason="This test needs a template not owned by sample_service.", run=False)
+def test_v3_process_notification_template_owner_mismatch(
+    notify_db_session, mocker, sample_service, sample_template, sample_template_without_email_permission
+):
+    """
+    Call the task with request data for a template the service doesn't own.
+    """
+
+    assert sample_template.template_type == SMS_TYPE
+    assert sample_template.service_id == sample_service.id
+
+    assert sample_template_without_email_permission.template_type == EMAIL_TYPE
+    assert sample_template_without_email_permission.service_id != sample_service.id
+    assert False
+
+    request_data = {
+        "id": str(uuid4()),
+        "notification_type": SMS_TYPE,
+        "phone_number": "+18006982411",
+        "template_id": sample_template.id,
+    }
+
+    v3_send_sms_notification_mock = mocker.patch("app.celery.v3.notification_tasks.v3_send_sms_notification.delay")
+    v3_process_notification(request_data, sample_service.id, None, KEY_TYPE_TEST)
+    v3_send_sms_notification_mock.assert_not_called()
+
+    query = select(Notification).where(Notification.id == request_data["id"])
+    notification = notify_db_session.session.execute(query).one().Notification
+    assert notification.status == NOTIFICATION_PERMANENT_FAILURE
+    assert notification.status_reason == "The service does not own the template."
+
+
+def test_v3_process_notification_template_type_mismatch_1(notify_db_session, mocker, sample_service, sample_template):
+    """
+    Call the task with request data for an e-mail notification, but specify an SMS template.
+    """
+
+    assert sample_template.template_type == SMS_TYPE
+
+    request_data = {
+        "id": str(uuid4()),
+        "notification_type": EMAIL_TYPE,
+        "email_address": "test@va.gov",
+        "template_id": sample_template.id,
+    }
+
+    v3_send_email_notification_mock = mocker.patch("app.celery.v3.notification_tasks.v3_send_email_notification.delay")
+    v3_process_notification(request_data, sample_service.id, None, KEY_TYPE_TEST)
+    v3_send_email_notification_mock.assert_not_called()
+
+    query = select(Notification).where(Notification.id == request_data["id"])
+    notification = notify_db_session.session.execute(query).one().Notification
+    assert notification.status == NOTIFICATION_PERMANENT_FAILURE
+    assert notification.status_reason == "The template type does not match the notification type."
+
+
+def test_v3_process_notification_template_type_mismatch_2(
+    notify_db_session, mocker, sample_service, sample_email_template
+):
+    """
+    Call the task with request data for an SMS notification, but specify an e-mail template.
+    """
```

Comment:
Is this an effective unit test, given that integrated, sending an SMS with an email template fails?

Reply:
It's effective in that it tests the logic I wrote.  The failure, which we have captured with a spike ticket, is outside of the v3 code.